docs/source/working_with_aiida/caching.rst

```diff
@@ -65,88 +113,69 @@ This means that caching is enabled for ``TemplatereplacerCalculation`` and ``Str
 
-.. code:: python
 
-    from aiida.engine import workfunction
-    from aiida.orm import Int
+.. _caching_limitations:
 
-    @workfunction
-    def select(a, b):
-        return b
+Limitations
+-----------
 
-    d = Int(1)
-    r1 = select(d, d)
-    r2 = select(Int(1), Int(1))
+1. The current implementation of caching for data nodes clones not only the graph representation of the reused node but also the underlying data in the database and file repository. 
+   This could be improved in order to reduce data duplication.
 
-The two ``select`` workfunctions have the same inputs as far as their hashes go. However, the first call uses the same input node twice, while the second one has two different inputs. If the second call should be cached from the first one, it is not clear which of the two input nodes should be returned.
+2. The caching mechanism for calculations *should* trigger only when the inputs and the calculation to be performed are exactly the same.  
+   Edge cases where this assumption might be violated include cases where the calculation parser is in a different python module than the calculation and the developer made changes without updating the version number of the plugin.
 
-While this example might seem contrived, the conclusion is valid more generally: Because workchains can return nodes from their history, they cannot be cached. Since even two equivalent workchains (with the same inputs) can have a different history, there is no way to deduce which links should be created on the new workchain without actually running it.
+3. The constraint that the shape of the provenance graph should be independent of whether caching is enabled or not, imposes limitations on the possible caching operations.
+   In order to reuse a cached node, we need to know not only its contents but also how the new node should be linked to its parents and children.
 
-Overall, this limitation is acceptable: The runtime of AiiDA workchains is usually dominated by time spent inside expensive calculations. Since these can be avoided with the caching mechanism, it still improves the runtime and required computer resources a lot.
+   * **Data nodes:** Making a copy of a data node should not change its links, so AiiDA just needs to link the new node to the direct ancestors of the old node.
+   * **Calculations nodes:** The node can have inputs and creates new data nodes as outputs. Again, the new node needs to replicate these links. In order to make it look as if the calculation had produced its own outputs, the output nodes are copied and linked as well.
+   * **Workflows nodes:** Workflows differ from calculations in that they can *return* an input node or an output node created by a calculation. 
+     This is not compatible with the cloning model chosen for calculations (leave input nodes untouched, clone output nodes). For this reason, workflows are not cached.
```

Comment:
I understand that you wanted to simplify the reasoning why `WorkflowNodes` are not cached, but I think this makes it worse. Sure the sentence here is less complicated, but it does not explain the *why* at all. I would really keep the example that demonstrates the problem, it is really important

Reply:
I changed this, since I felt the example given did not actually illustrate the problem.
Let's rediscuss this in person, I will then adapt the text accordingly.

Reply:
I've updated this section, and it does now give a hint as to why caching workflows is not straightforward.